Introduce location tracking in the tokenizer and parser

ankrgyl opened the pull request:
This revives PR #288 and #514, rebased against the latest and done with fewer variable renames (limiting merge conflicts).

Closes https://github.com/sqlparser-rs/sqlparser-rs/issues/524

ankrgyl commented:
@alamb @AugustoFKL this PR addresses #524.

This change is reasonably contained, but one thing that could make it even smaller would be maintaining the interface (i.e.`next_token()` returns a `Token` not a `TokenWithLocation`) and maybe adding something like `last_location()` to access the location just when you need it (e.g. for error messages).

alamb commented:
Thanks @ankrgyl  -- I'll try to look at this PR more carefully later this week

AugustoFKL commented:
@ankrgyl @alamb sorry for the delay. This has been a hectic week for me. I'm on vacation next week, and will be able to review this properly

alamb commented on src/tokenizer.rs:
```diff
@@ -300,6 +300,41 @@ impl fmt::Display for Whitespace {
+    pub column: u64,
+}
+
+/// A [Token] with [Location] attached to it
```
We can probably avoid many of the changes in this PR by implementing `PartialEq` between Token and TokenWithLocation:

```rust
impl PartialEq<Token>` for TokenWithLocation
```

Which would allow things like

```rust
        if self.peek_nth_token(1).token == Token::RArrow {
```

to keep working

alamb commented on src/tokenizer.rs:
```diff
@@ -321,58 +356,88 @@ impl fmt::Display for TokenizerError {
 #[cfg(feature = "std")]
 impl std::error::Error for TokenizerError {}
 
+struct State<'a> {
```
👍

alamb reviewed (commented):
This PR looks great @ankrgyl  -- thank you for proposing it; I apologize for the delay in finding time to review it

I think we can avoid many of the changes with an `impl PartialEq` as I mentioned

Otherwise, all this PR needs is some tests and I think it would be ready to merge.

Thanks again!

ankrgyl commented:
@alamb updated with the `PartialEq` suggestion (great idea!). For tests -- do you have a suggestion on what would be best? I can tweak a few of the tests to tokenize w/ location and hardcode them in, or do something else?

alamb commented:
> For tests -- do you have a suggestion on what would be best? I can tweak a few of the tests to tokenize w/ location and hardcode them in, or do something else?

I think the goal should be that if we accidentally break the location parsing code some tests will fail. I think ensuring the locations are set on parse errors would be particularly helpful. 

So perhaps you could add a few tests that assert locations of parsed tokens, and then a test that has a parsing error that retrieves the location of the error (if this last test needs more code changes, we could do it as a follow on PR)


Thank you so much for pushing this along @ankrgyl  -- this is going to be a great addition (and is a long asked for feature)

ankrgyl commented:
Added a few tests. We'll need to thread parser locations through as a follow up PR, but I added a test that fails in the parser due to a tokenizer error (and asserts the location) + a test that we can fill in with locations once the parser carries along that info.

alamb commented on src/parser.rs:
```diff
@@ -6592,4 +6592,31 @@ mod tests {
+        assert_eq!(
+            ast,
+            Err(ParserError::TokenizerError(
+                "Unterminated string literal at Line: 1, Column 5".to_string()
```
❤️

alamb reviewed (approved):
I think this looks great -- thank you @ankrgyl

Would you like a chance to review @AugustoFKL ?

alamb commented:
This appears to have a logical conflict with master.

I am planning on cutting a release soon -- maybe given the nature of this change I will make release right before this change, and then merge this one. That way people who use the tip of this repository can effectively beta test the changes before we make a release for everyone 🤔

ankrgyl commented:
That seems wise to me. We may even want to flesh out the (potentially breaking) parser changes so that users who upgrade only need to adjust to breakages once. Happy to contribute those and collaborate with you on that.

ankrgyl commented:
Additionally, I just rebased and updated the merge conflicts.

alamb commented:
> That seems wise to me. We may even want to flesh out the (potentially breaking) parser changes so that users who upgrade only need to adjust to breakages once. Happy to contribute those and collaborate with you on that.


That would be awesome. I plan to make a release soon (TM) -- either later today or over the weekend sometime. Then let's merge this one in and iterate.

alamb commented:
@ankrgyl  I took the liberty of fixing the clippy CI error and merging up from main to this branch.

alamb commented:
>  That seems wise to me. We may even want to flesh out the (potentially breaking) parser changes so that users who upgrade only need to adjust to breakages once. Happy to contribute those and collaborate with you on that.


This is merged and should be included in the 0.29.0 release (eta in approximately 1 months time) I would love to start collaborating to fixup the parser. The first step might be to list / sketch out the changes that are needed.

Thanks again -- this is super exciting!
